engines/dfs: add DAOS File System (dfs) engine

johannlombardi opened the pull request:
DAOS is a new scale-out open-source object store.
See https://github.com/daos-stack/daos for more information.

This patch adds a new fio engine to support the filesystem layer built
on top of DAOS called DFS (DAOS File System). It supports asynchronous
read/write operations.

Signed-off-by: Johann Lombardi <johann.lombardi@intel.com>
Co-authored-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>
Co-authored-by: Vishwanath Venkatesan <vishwanath.venkatesan@intel.com>
Co-authored-by: Ioannis Galanis <ioannis.galanis@intel.com>

sitsofe commented on engines/dfs.c:
```diff
@@ -0,0 +1,564 @@
+		return EINVAL;
+	}
+
+	/** Retrieve object class to use */
```
Why two starts?

johannlombardi replied:
Not sure I get this one. The code below just parses the object_class option (if any) and convert it from a string to our object class identifier enum. Could you please elaborate?

sitsofe replied:
Totally trivial: "/**" <- there's a second star :-)

johannlombardi replied:
ah, got it. Just different coding style in the DAOS project. Anyway, fixed.

sitsofe commented on engines/dfs.c:
```diff
@@ -0,0 +1,564 @@
+	rc = daos_fini();
+	if (rc)
+		log_err("failed to finalize daos: %d\n", rc);
+}
```
Don't you want to return/set an error so callers know something went wrong?

johannlombardi replied:
Good point. fixed.

sitsofe reviewed (changes requested):
Some minor comments after a brief first pass review.

sitsofe commented:
Oh and don't forget to add a prefix to your commit's title (see https://github.com/axboe/fio/commits/master/engines for ideas).

johannlombardi commented on configure:
```diff
@@ -2179,6 +2182,32 @@ if test "$libnbd" != "no" ; then
 
+##########################################
+# check for dfs (DAOS File System)
+if test "$dfs" != "yes" ; then
```
fixed.

johannlombardi commented on engines/dfs.c:
```diff
@@ -0,0 +1,564 @@
+};
+
+static int
+daos_fio_global_init(struct thread_data *td)
```
fixed all occurrences.

johannlombardi commented:
@sitsofe Thanks a lot for the review. I have addressed all your points except the one related to the two starts that I don't grok.
I have also updated the title of the PR to include a prefix. Let me know how it looks.

sitsofe commented on engines/dfs.c:
```diff
@@ -0,0 +1,571 @@
+
+	pthread_mutex_lock(&daos_mutex);
+
+	/* Allocate space for DAOS-related data */
```
I think you can skip this comment :-)

johannlombardi replied:
done.

sitsofe commented on configure:
```diff
@@ -2179,6 +2183,33 @@ if test "$libnbd" != "no" ; then
+  daos_handle_t	coh;
+  dfs_t		*dfs;
+
+  (void)dfs_mount(poh, coh, O_RDWR, &dfs);
```
Fio generally a space after the bracket in this situation.

johannlombardi replied:
done

sitsofe commented on engines/dfs.c:
```diff
@@ -0,0 +1,571 @@
+		if (dd && dd->io_us)
+			free(dd->io_us);
+		if (dd)
+			free(dd);
```
Did you know that `free()` "works" on `NULL` as well (I've only recently come across https://stackoverflow.com/a/6084233 )?

johannlombardi replied:
I do. Let me remove the condition then if you think that it makes the code more readable.

sitsofe commented on engines/dfs.c:
```diff
@@ -0,0 +1,571 @@
+		td_verror(td, rc, "daos_eq_destroy");
+	}
+
+	free(dd->io_us);
```
Do you know `dd` must not be NULL?

johannlombardi replied:
Yes, it is tested at the beginning of this function.

sitsofe reviewed (commented):
Looking good - only trivial comments from me. Can you squash your commits together so everything arrives as one?

johannlombardi commented:
@sitsofe Patch updated as per your comments and all commits squashed together. Thanks.

johannlombardi commented:
Just fixed a minor issue reported by a user. No further changes expected as far as I am concerned.

johannlombardi commented:
@axboe Any chance to consider this PR for review/integration? Thanks in advance.

johannlombardi commented:
I have added a fix for the compilation issue on msys2 and also tested the new dfs engine with different configurations on a sizable DAOS system. @axboe any chance to review/land this PR? Thanks in advance.

axboe commented:
I'll take a look today.

axboe commented:
Looks pretty clean to me. The only thing I noticed is that the daos_mutex is used a bit liberally for init, there should be no need to grab it before allocating and initializing a new data_data structure. It should only be needed once you get to checking for the init point, incrementing num_threads, etc. But that's minor.

johannlombardi commented:
Thanks a lot @axboe. I will look into the mutex thing.
